include/mesh_multiphase.tcc

```diff
@@ -0,0 +1,494 @@
+          std::bind(&mpm::NodeBase<Tdim>::volume, std::placeholders::_1,
+                    mpm::ParticlePhase::SinglePhase),
+          std::bind(&mpm::NodeBase<Tdim>::update_volume, std::placeholders::_1,
+                    false, mpm::ParticlePhase::SinglePhase,
+                    std::placeholders::_2));
+    }
+#endif
+
+    // Compute nodal value of mass density
+    this->iterate_over_nodes_predicate(
+        std::bind(&mpm::NodeBase<Tdim>::compute_density, std::placeholders::_1),
+        std::bind(&mpm::NodeBase<Tdim>::status, std::placeholders::_1));
+
+    // Evaluate free surface particles
+    for (auto pitr = this->particles_.cbegin(); pitr != this->particles_.cend();
+         ++pitr) {
+      bool status = (*pitr)->compute_free_surface_by_density();
+      if (status) {
+        (*pitr)->assign_free_surface(status);
+      }
+    }
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+  }
+  return status;
+}
+
+//! Get free surface node set
+template <unsigned Tdim>
+std::set<mpm::Index> mpm::Mesh<Tdim>::free_surface_nodes() {
+  std::set<mpm::Index> id_set;
+  for (auto nitr = this->nodes_.cbegin(); nitr != this->nodes_.cend(); ++nitr)
+    if ((*nitr)->free_surface()) id_set.insert((*nitr)->id());
+  return id_set;
+}
+
+//! Get free surface cell set
+template <unsigned Tdim>
+std::set<mpm::Index> mpm::Mesh<Tdim>::free_surface_cells() {
+  std::set<mpm::Index> id_set;
+  for (auto citr = this->cells_.cbegin(); citr != this->cells_.cend(); ++citr)
+    if ((*citr)->free_surface()) id_set.insert((*citr)->id());
+  return id_set;
+}
+
+//! Get free surface particle set
+template <unsigned Tdim>
+std::set<mpm::Index> mpm::Mesh<Tdim>::free_surface_particles() {
+  std::set<mpm::Index> id_set;
+  for (auto pitr = this->particles_.cbegin(); pitr != this->particles_.cend();
+       ++pitr)
+    if ((*pitr)->free_surface()) id_set.insert((*pitr)->id());
+  return id_set;
+}
+
+//! Assign particle pore pressures
+template <unsigned Tdim>
+bool mpm::Mesh<Tdim>::assign_particles_pore_pressures(
+    const std::vector<std::tuple<mpm::Index, double>>&
+        particle_pore_pressures) {
+  bool status = true;
+
+  try {
+    if (!particles_.size())
+      throw std::runtime_error(
+          "No particles have been assigned in mesh, cannot assign pore "
+          "pressures");
+
+    for (const auto& particle_pore_pressure : particle_pore_pressures) {
+      // Particle id
+      mpm::Index pid = std::get<0>(particle_pore_pressure);
+      // Pore pressure
+      double pore_pressure = std::get<1>(particle_pore_pressure);
+
+      if (map_particles_.find(pid) != map_particles_.end())
+        map_particles_[pid]->assign_pressure(pore_pressure,
+                                             mpm::ParticlePhase::Liquid);
+    }
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
```

Comment:
Are we OK if this fails and we can continue the calculation with default values?

Reply:
I think that would be better if we stop the simulation.